### PR TITLE
feat(flat-table): flat table enhancements (FE-4065)

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour
Add custom row background color prop
Add alternative header color prop
Add horizontal border size control
Add vertical border size control

### Checklist

- [x] Commits follow our style guide
~~- [ ] Screenshots are included in the PR if useful~~
~~- [ ] All themes are supported if required~~
- [x] Unit tests added or updated if required
~~- [ ] Cypress automation tests added or updated if required~~
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
1. npm start
2. open:
http://localhost:9001/?path=/story/design-system-flat-table--with-custom-row-background-color
http://localhost:9001/?path=/story/design-system-flat-table--with-custom-horizontal-border-size
http://localhost:9001/?path=/story/design-system-flat-table--with-custom-vertical-borders
http://localhost:9001/?path=/story/design-system-flat-table--with-alternative-header-background

### CODE REVIEW

File: src/components/flat-table/flat-table-cell/flat-table-cell.d.ts [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import * as React from "react";
 import { SpaceProps } from "styled-system";
+import { TableBorderSize } from "..";
```
Comment (edleeks87) [inline src/components/flat-table/flat-table-cell/flat-table-cell.d.ts:3]: Do you need this import if it's declared locally below?

File: src/components/flat-table/flat-table-header/flat-table-header.spec.js [hunk 2]
```diff
@@ -22,4 +22,40 @@ describe("FlatTableHeader", () => {
       { modifier: "&&& > div" }
     );
   });
+
+  describe('with the "bgColor" also provided', () => {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table-header/flat-table-header.spec.js:26]: This description/ prop name doesn't seem to fit what's happening in the test

File: src/components/flat-table/flat-table-row/flat-table-row.style.js [hunk 3]
```diff
@@ -55,6 +60,16 @@ const StyledFlatTableRow = styled.tr`
   table-layout: fixed;
   width: auto;
 
+  ${StyledFlatTableCell} {
+    ${({ bgColor }) => bgColor && `background-color: ${bgColor};`}
```
Comment (edleeks87) [inline src/components/flat-table/flat-table-row/flat-table-row.style.js:64]: If we use the `toColor` util we could allow consumer's to use the palette as well as hex and css strings

https://github.com/Sage/carbon/blob/b1e510427d75bc191e35467f85ec93995f0ef8a3/src/style/utils/color.js

File: src/components/flat-table/flat-table-row/flat-table-row.style.js [hunk 4]
```diff
@@ -55,6 +60,16 @@ const StyledFlatTableRow = styled.tr`
   table-layout: fixed;
   width: auto;
 
+  ${StyledFlatTableCell} {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table-row/flat-table-row.style.js:63]: I think you'll need to target `FlatTableCheckbox` and `FlatTableRowHeader` here as well

image

**Review by edleeks87 — COMMENTED** [review]
Is there a way of overriding the colour of the borders as it's listed on the ticket?

**Review by edleeks87 — DISMISSED** [review]

**Review by nicktitchmarsh — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Comment by harpalsingh** [comment]
Hey @grabkowski just one minor thing, the border color of header cells should remain the same, you can see from this zoomed visual and hex codes the border used with alternative header is different

Screenshot 2021-06-03 at 12 47 03

**Review by edleeks87 — DISMISSED** [review]

**Review by nicktitchmarsh — DISMISSED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Review by sianford — APPROVED** [review]

**Comment by sianford** [comment]
Tested in Safari, Chrome, Firefox and Edge. Testing passed ✅

**Pull request merged**